[DeviceRTL][AMDGPU] Split atmi_memcpy for d2h and h2d

pdhaliwal-amd opened the pull request:
This patch is mostly NFC as the underlying copy is still done by generic atmi_memcpy. Call sites, however, are updated to use the respective copy method as they already know which direction they are going to copy in.  Future patch will remove the check for the direction from the memcpy implementation.

JonChesterfield commented on openmp/libomptarget/plugins/amdgpu/impl/atmi.cpp:
```diff
@@ -38,6 +38,16 @@ atmi_status_t atmi_memcpy(hsa_signal_t sig, void *dest, const void *src,
   return core::Runtime::Memcpy(sig, dest, src, size);
 }
 
+atmi_status_t atmi_memcpy_h2d(hsa_signal_t sig, void *host_dest,
```
parameter names are backwards. h2d as in 'host to device' copies from host to device, so host_src and device_dst. Though src/dst seems fine too.

JonChesterfield commented on openmp/libomptarget/plugins/amdgpu/impl/atmi_runtime.h:
```diff
@@ -182,6 +182,16 @@ atmi_status_t atmi_free(void *ptr);
 atmi_status_t atmi_memcpy(hsa_signal_t sig, void *dest, const void *src,
```
drop atmi_memcpy from header, helps catch call sites that have been missed from the patch

JonChesterfield commented on openmp/libomptarget/plugins/amdgpu/src/rtl.cpp:
```diff
@@ -353,23 +353,42 @@ class RTLDeviceInfoTy {
-  atmi_status_t freesignalpool_memcpy(void *dest, const void *src,
-                                      size_t size) {
+  atmi_status_t freesignalpool_memcpy_h2d(void *dest, const void *src,
+                                          size_t size) {
```
The hsa_memory_copy hack should be in atmi_memcpy, as then it's in one place instead of four.

The two implementations are the same, other than the atmi_memcpy_ call, so can implement as one function that takes a pointer. Typing directly into this comment box so syntax may be slightly off:
```
template <typename F> // or use an explicitly typed function pointer
static atmi_status_t freesignalpool_memcpy_impl(void *dest, const void *src, size_t size, F func) {	

hsa_signal_t s = FreeSignalPool.pop();
if (s.handle == 0) {
  return ATMI_STATUS_ERROR;
}
atmi_status_t r = func(s, dest, src, size);
FreeSignalPool.push(s);
return r;
}

atmi_status_t freesignalpool_memcpy_h2d(void *dest, const void *src, size_t size) {	
  return freesignalpool_memcpy_impl(dest, src, size, atmi_memcpy_h2d);
}
atmi_status_t freesignalpool_memcpy_d2h(void *dest, const void *src, size_t size) {	
  return freesignalpool_memcpy_impl(dest, src, size, atmi_memcpy_d2h);
}
```

JonChesterfield reviewed (changes requested):
Broadly reasonable, though more code than necessary.

If you look at the start of Runtime::Memcpy, which atmi_memcpy wraps, you'll find the current logic for guessing the direction of copy. We can use that to check the directions are consistent, but that may as well be a second patch.

pdhaliwal-amd commented:
Updated with requested changes. I had seen atmi_memcpy_no_signal being used only in one place hostrpc_handler_SERVICE_DEMO method, so I had moved its definition for hostrpc_execute_service.

JonChesterfield commented on openmp/libomptarget/plugins/amdgpu/src/rtl.cpp:
```diff
@@ -353,27 +354,31 @@ class RTLDeviceInfoTy {
   static const int Default_WG_Size =
       llvm::omp::AMDGPUGpuGridValues[llvm::omp::GVIDX::GV_Default_WG_Size];
 
+  using MemcpyFunc = std::function<atmi_status_t (hsa_signal_t,
```
Generally in favour of std::function, though it does induce some runtime overhead. As part of making this library exception proof (which is some way off), I'd prefer we go with the raw function pointer `using MemcpyFunc = atmi_status_t (*)(hsa_signal_t, void*,const void*, size_t)` or similar.

JonChesterfield commented:
LG, thanks. Please squash the three commits into one before committing.
